File: ios/Classes/ApptentiveFlutterPlugin.m
#import "ApptentiveFlutterPlugin.h"

#import "ApptentiveMain.h"

inline static _Nullable id fromNullable(_Nullable id value) {
  return [value isKindOfClass:[NSNull class]] ? nil : value;
}

static BOOL parseLogLevel(NSString *value, ApptentiveLogLevel *outResult) {
  if ([@"LogLevel.verbose" isEqualToString:value]) {
    if (outResult) *outResult = ApptentiveLogLevelVerbose;
    return YES;
  }
  if ([@"LogLevel.debug" isEqualToString:value]) {
    if (outResult) *outResult = ApptentiveLogLevelDebug;
    return YES;
  }
  if ([@"LogLevel.info" isEqualToString:value]) {
    if (outResult) *outResult = ApptentiveLogLevelInfo;
    return YES;
  }
  if ([@"LogLevel.warn" isEqualToString:value]) {
    if (outResult) *outResult = ApptentiveLogLevelWarn;
    return YES;
  }
  if ([@"LogLevel.error" isEqualToString:value]) {
    if (outResult) *outResult = ApptentiveLogLevelError;
    return YES;
  }

  return NO;
}

static ApptentiveConfiguration *unpackConfiguration(NSDictionary *info) {
  NSString *apptentiveKey = info[@"key"];
  NSString *apptentiveSignature = info[@"signature"];
  ApptentiveConfiguration *configuration = [ApptentiveConfiguration configurationWithApptentiveKey:apptentiveKey apptentiveSignature:apptentiveSignature];

  // log level
  ApptentiveLogLevel logLevel;
  NSString *logLevelValue = fromNullable(info[@"log_level"]);
  if (logLevelValue != nil) {
    if (parseLogLevel(logLevelValue, &logLevel)) {
      configuration.logLevel = logLevel;
    } else {
      NSLog(@"Apptentive: Unknown log level: %@", logLevelValue);
    }
  }

  // Sanitize Log Messages
  id shouldSanitizeLogMessages = fromNullable(info[@"should_sanitize_log_messages"]);
  if (shouldSanitizeLogMessages != nil) {
    configuration.shouldSanitizeLogMessages = [shouldSanitizeLogMessages boolValue];
  }

  // Show Info Button
  id shouldShowInfoButton = fromNullable(info[@"should_show_info_button"]);
  if (shouldShowInfoButton != nil) {
    configuration.showInfoButton = [shouldShowInfoButton boolValue];
  }

  // Enable Debug Log File
  id enableDebugLogFile = fromNullable(info[@"enable_debug_log_file"]);
  if (enableDebugLogFile != nil) {
    configuration.enableDebugLogFile = [enableDebugLogFile boolValue];
  }

  // Gather Carrier Info
  id gatherCarrierInfo = fromNullable(info[@"gather_carrier_info"]);
  if (gatherCarrierInfo != nil) {
    configuration.gatherCarrierInfo = [gatherCarrierInfo boolValue];
  }

  // Terms and conditions
  id termsAndConditionsPacked = fromNullable(info[@"terms_and_conditions"]);
  if (termsAndConditionsPacked != nil) {
    // Unpack
    NSString *bodyText = fromNullable(termsAndConditionsPacked[@"body_text"]);
    NSString *linkText = fromNullable(termsAndConditionsPacked[@"link_text"]);
    NSURL *linkUrl = fromNullable(termsAndConditionsPacked[@"link_url"]);
    TermsAndConditions *termsAndConditions = [[TermsAndConditions alloc] initWithBodyText:bodyText linkText:linkText linkURL:linkUrl];
    configuration.surveyTermsAndConditions = termsAndConditions;
  }

  // Set distribution information
  configuration.distributionName = @"Flutter";
  configuration.distributionVersion = @"5.7.1-rc.6";

  return configuration;
}

@interface NSNumber (ApptentiveBoolean)

- (BOOL)apptentive_isBoolean;

@end

@implementation NSNumber (ApptentiveBoolean)

- (BOOL)apptentive_isBoolean {
    CFTypeID boolID = CFBooleanGetTypeID();
    CFTypeID numID = CFGetTypeID((__bridge CFTypeRef)(self));
    return numID == boolID;
}

@end

@interface ApptentiveFlutterPlugin ()

@property (strong, nonatomic) FlutterMethodChannel* channel;
@property (strong, nonatomic) NSData* deviceToken;

@end

@implementation ApptentiveFlutterPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel *channel = [FlutterMethodChannel
      methodChannelWithName:@"apptentive_flutter"
            binaryMessenger:[registrar messenger]];
  ApptentiveFlutterPlugin* instance = [[ApptentiveFlutterPlugin alloc] initWithChannel:channel];
  [registrar addApplicationDelegate:instance];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (instancetype)initWithChannel:(FlutterMethodChannel *)channel {
  self = [super init];

  if (self) {
    _channel = channel;
    // Register to grab new device/push token
    [[UIApplication sharedApplication] registerForRemoteNotifications];
  }

  return self;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"register" isEqualToString:call.method]) {
    [self handleRegisterCall:call result: result];
  } else if ([@"showMessageCenter" isEqualToString:call.method]) {
    [self handleShowMessageCenterCall:call result: result];
  } else if ([@"engage" isEqualToString:call.method]) {
    [self handleEngageCall:call result: result];
  } else if ([@"canShowInteraction" isEqualToString:call.method]) {
    [self handleCanShowInteractionCall:call result: result];
  } else if ([@"setPersonName" isEqualToString:call.method]) {
    [self handleSetPersonNameCall:call result: result];
  } else if ([@"setPersonEmail" isEqualToString:call.method]) {
    [self handleSetPersonEmailCall:call result: result];
  } else if ([@"addCustomPersonData" isEqualToString:call.method]) {
    [self handleAddCustomPersonDataCall:call result: result];
  } else if ([@"removeCustomPersonData" isEqualToString:call.method]) {
    [self handleRemoveCustomPersonDataCall:call result: result];
  } else if ([@"addCustomDeviceData" isEqualToString:call.method]) {
    [self handleAddCustomDeviceDataCall:call result: result];
  } else if ([@"removeCustomDeviceData" isEqualToString:call.method]) {
    [self handleRemoveCustomDeviceDataCall:call result: result];
  } else if ([@"login" isEqualToString:call.method]) {
    [self handleLoginCall:call result: result];
  } else if ([@"logout" isEqualToString:call.method]) {
    [self handleLogoutCall:call result: result];
  } else if ([@"setPushNotificationIntegration" isEqualToString:call.method]) {
    [self handleSetPushNotificationIntegrationCall:call result: result];
  } else if ([@"getUnreadMessageCount" isEqualToString:call.method]) {
    [self handleGetUnreadMessageCount:call result: result];
  } else if ([@"registerListeners" isEqualToString:call.method]) {
    [self handleRegisterListeners:call result: result];
  } else if ([@"requestPushPermissions" isEqualToString:call.method]) {
    [self handleRequestPushPermissions:call result: result];
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (void)handleRegisterCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  ApptentiveConfiguration *configuration = unpackConfiguration(call.arguments[@"configuration"]);
  [Apptentive registerWithConfiguration:configuration];
  result(@YES);
}

- (void)handleShowMessageCenterCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSDictionary *customData = call.arguments;
  if (![self isRegistered]){
    NSLog(@"Apptentive is not initialized, cannot show Message Center.");
    result(@NO);
    return;
  }
  [Apptentive.shared presentMessageCenterFromViewController:nil
                                             withCustomData:customData
                                                 completion:^(BOOL presented) {
    result([NSNumber numberWithBool:presented]);
  }];
}

// Set notification observers
- (void)handleRegisterListeners:(FlutterMethodCall*)call result:(FlutterResult)result {
  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(messageCenterUnreadCountChangedNotification:) name:ApptentiveMessageCenterUnreadCountChangedNotification object:nil];
  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(surveySentNotification:) name:ApptentiveSurveySentNotification object:nil];
  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(surveyCancelledNotification:) name:ApptentiveSurveyCancelledNotification object:nil];
  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(messageSentNotification:) name:ApptentiveMessageSentNotification object:nil];

  [Apptentive.shared setAuthenticationFailureCallback:^void (ApptentiveAuthenticationFailureReason reason, NSString *errorMessage) {
    [self.channel invokeMethod:@"onAuthenticationFailed"
          arguments:@{
            @"reason": fromNullable(@(reason)),
            @"errorMessage": fromNullable(errorMessage),
          }
    ];
  }];
}

- (void)handleEngageCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString *event = call.arguments[@"event_name"];
  NSDictionary *customData = fromNullable(call.arguments[@"custom_data"]);
  if (![self isRegistered]){
    NSLog(@"Apptentive is not initialized, cannot engage event: %@", event);
    result(@NO);
    return;
  }
  [Apptentive.shared engage:event
             withCustomData:customData
         fromViewController:nil
                 completion:^(BOOL engaged) {
    result([NSNumber numberWithBool:engaged]);
  }];
}

- (void)handleCanShowInteractionCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString *event = call.arguments[@"event_name"];
  if (![self isRegistered]){
    NSLog(@"Apptentive is not initialized, cannot show any interactions for event: %@", event);
    result(@NO);
    return;
  }
  [Apptentive.shared queryCanShowInteractionForEvent:event completion:^(BOOL canShowInteraction) {
      result([NSNumber numberWithBool:canShowInteraction]);
  }];
}

- (void)handleSetPersonNameCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if (![self isRegistered]){
    NSLog(@"Apptentive is not initialized, cannot set person name.");
    result(@NO);
    return;
  }
  NSString *name = call.arguments[@"name"];
  [Apptentive.shared setPersonName:name];
  result(@YES);
}

- (void)handleSetPersonEmailCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if (![self isRegistered]){
    NSLog(@"Apptentive is not initialized, cannot set person email.");
    result(@NO);
    return;
  }
  NSString *email = call.arguments[@"email"];
  [Apptentive.shared setPersonEmailAddress:email];
  result(@YES);
}

- (void)handleAddCustomPersonDataCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString *key = call.arguments[@"key"];
  id value = fromNullable(call.arguments[@"value"]);
  if ([value isKindOfClass:[NSString class]] || value == nil) {
    [Apptentive.shared addCustomPersonDataString:value withKey:key];
    result(@YES);
  } else if ([value isKindOfClass:[NSNumber class]]) {
    if ([value apptentive_isBoolean]) {
      [Apptentive.shared addCustomPersonDataBool:[value boolValue] withKey:key];
    } else {
      [Apptentive.shared addCustomPersonDataNumber:value withKey:key];
    }
    result(@YES);
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (void)handleRemoveCustomPersonDataCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString *key = call.arguments[@"key"];
  [Apptentive.shared removeCustomPersonDataWithKey:key];
  result(@YES);
}

- (void)handleAddCustomDeviceDataCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString *key = call.arguments[@"key"];
  id value = fromNullable(call.arguments[@"value"]);
  if ([value isKindOfClass:[NSString class]] || value == nil) {
    [Apptentive.shared addCustomDeviceDataString:value withKey:key];
    result(@YES);
  } else if ([value isKindOfClass:[NSNumber class]]) {
    if ([value apptentive_isBoolean]) {
      [Apptentive.shared addCustomDeviceDataBool:[value boolValue] withKey:key];
    } else {
      [Apptentive.shared addCustomDeviceDataNumber:value withKey:key];
    }
    result(@YES);
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (void)handleRemoveCustomDeviceDataCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString *key = call.arguments[@"key"];
  [Apptentive.shared removeCustomDeviceDataWithKey:key];
  result(@YES);
}

- (void)handleLoginCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString *token = call.arguments[@"token"];
  [Apptentive.shared logInWithToken:token
                         completion:^(BOOL success, NSError * _Nonnull error) {
    result([NSNumber numberWithBool:success]);
  }];
}

- (void)handleLogoutCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  [Apptentive.shared logOut];
  result(@YES);
}

- (void)handleSetPushNotificationIntegrationCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  // Convert the push provider into an integer using the util method
  NSInteger pushProvider = [self parsePushProvider:call.arguments[@"push_provider"]];
  // Call native register method with saved device token if token is not null
  if (_deviceToken != NULL) {
    [Apptentive.shared setPushNotificationIntegration:pushProvider withDeviceToken:_deviceToken];
    result(@YES);
  } else {
    result(@NO);
  }
}

- (NSInteger)parsePushProvider:(NSString*) pushProvider {
  if ([pushProvider containsString:@"apptentive"]) {
    return ApptentivePushProviderApptentive;
  }
  if ([pushProvider containsString:@"amazon"]) {
    return ApptentivePushProviderAmazonSNS;
  }
  if ([pushProvider containsString:@"parse"]) {
    return ApptentivePushProviderParse;
  }
  if ([pushProvider containsString:@"urban_airship"]) {
    return ApptentivePushProviderUrbanAirship;
  }
  [NSException raise:@"Apptentive Error: Unknown push provider" format:@"Push provider %@ is invalid", pushProvider];
  return -1;
}

- (void)handleGetUnreadMessageCount:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSNumber* count = [NSNumber numberWithInteger:Apptentive.shared.unreadMessageCount];
  result(count);
}

- (void)handleRequestPushPermissions:(FlutterMethodCall*)call result:(FlutterResult)result {
  UNUserNotificationCenter* center = [UNUserNotificationCenter currentNotificationCenter];
  [center getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings* _Nonnull settings) {
    switch (settings.authorizationStatus) {
      // End consumer has not seen permission dialog, show it
      case UNAuthorizationStatusNotDetermined: {
        center.delegate = self;
        [center requestAuthorizationWithOptions:(UNAuthorizationOptionAlert + UNAuthorizationOptionSound) completionHandler:^(BOOL granted, NSError * _Nullable error) {
          if (granted) {
            [[UIApplication sharedApplication] registerForRemoteNotifications];
          }
        }];
        break;
      }
      // End consumer already accepted, register for remote notifications
      case UNAuthorizationStatusAuthorized: {
        [[UIApplication sharedApplication] registerForRemoteNotifications];
      }
      default:
        break;
    }
  }];
}

- (void)getUnreadMessageCount:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSNumber* count = [NSNumber numberWithInteger:Apptentive.shared.unreadMessageCount];
  result(count);
}

- (void)getUnreadMessageCount:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSNumber* count = [NSNumber numberWithInteger:Apptentive.shared.unreadMessageCount];
  result(count);
}

// Notification Functions

- (void)messageCenterUnreadCountChangedNotification:(NSNotification *)notification {
  [self.channel invokeMethod:@"onUnreadMessageCountChanged"
        arguments:@{
          @"count" : fromNullable(notification.userInfo[@"count"]),
        }
  ];
}

- (void)surveySentNotification:(NSNotification *)notification {
  [self.channel invokeMethod:@"onSurveyFinished"
        arguments:@{
          @"completed" : @YES,
        }
  ];
}

- (void)surveyCancelledNotification:(NSNotification *)notification {
    [self.channel invokeMethod:@"onSurveyFinished"
          arguments:@{
            @"completed" : @NO,
          }
    ];
}

- (void)messageSentNotification:(NSNotification *)notification {
  NSString * sentByUser = notification.userInfo[@"sentByUser"];
    if(!sentByUser) {
        sentByUser = @"";
    }
  [self.channel invokeMethod:@"onMessageSent"
        arguments:@{
          @"sentByUser" : sentByUser,
        }
  ];
}

- (void)onAuthenticationFailed:(NSString *)reason errorMessage:(NSString *)errorMessage {
  [self.channel invokeMethod:@"onAuthenticationFailed"
        arguments:@{
          @"reason": reason,
          @"errorMessage": errorMessage,
        }
  ];
}

- (BOOL)isRegistered {
  return Apptentive.shared.apptentiveKey != nil && Apptentive.shared.apptentiveSignature != nil;
}

// Because Apptentive.shared is not the delegate, pass data into its internal functions for push handling
- (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)(void))completionHandler {
  [Apptentive.shared userNotificationCenter:center didReceiveNotificationResponse:response withCompletionHandler:completionHandler];
}

// Because Apptentive.shared is not the delegate, pass data into its internal functions for push handling
- (void)userNotificationCenter:(UNUserNotificationCenter *)center willPresentNotification:(UNNotification *)notification withCompletionHandler:(void (^)(UNNotificationPresentationOptions options))completionHandler {
  [Apptentive.shared userNotificationCenter:center willPresentNotification:notification withCompletionHandler:completionHandler];
}

// Save the device token after registering for remote notifications
- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken {
    self.deviceToken = deviceToken;
}

// Tell Native Apptentive to handle the notification
- (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo fetchCompletionHandler:(void (^)(UIBackgroundFetchResult result)) completionHandler {
  [Apptentive.shared didReceiveRemoteNotification:userInfo fetchCompletionHandler:completionHandler];
}

// Log the error if registering deviceCustomDataRemoverTester notifications failed
- (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)error {
  NSLog(@"Apptentive Error: Failed to register for remote notifications with error: %@", [error localizedDescription]);
}

@end

File: ios/Classes/ApptentiveFlutterPlugin.swift
import Flutter
import UIKit
import ApptentiveKit
import PhotosUI

public class ApptentiveFlutterPlugin: NSObject, FlutterApplicationLifeCycleDelegate, FlutterPlugin {

  private static let errorCode = "Apptentive Error"
  private var observation: NSKeyValueObservation?
  private let channel: FlutterMethodChannel

  // Register the method channel and plugin instance
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "apptentive_flutter", binaryMessenger: registrar.messenger())
    let instance = ApptentiveFlutterPlugin(channel: channel)
    registrar.addMethodCallDelegate(instance, channel: channel)
    registrar.addApplicationDelegate(instance)
  }

  // Handle the flutter method call, delegating it based on the method name
  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "register": handleRegisterCall(call, result)
    case "engage": handleEngageCall(call, result)
    case "showMessageCenter": handleShowMessageCenter(call, result)
    case "canShowMessageCenter": handleCanShowMessageCenterCall(call, result)
    case "getUnreadMessageCount": handleGetUnreadMessageCount(call, result)
    case "setPersonName": handleSetPersonNameCall(call, result)
    case "setPersonEmail": handleSetPersonEmailCall(call, result)
    case "addCustomPersonData": handleAddCustomPersonDataCall(call, result)
    case "removeCustomPersonData": handleRemoveCustomPersonDataCall(call, result)
    case "addCustomDeviceData": handleAddCustomDeviceDataCall(call, result)
    case "removeCustomDeviceData": handleRemoveCustomDeviceDataCall(call, result)
    case "canShowInteraction": handleCanShowInteractionCall(call, result)
    case "setPushNotificationIntegration": handleSetPushNotificationIntegrationCall(call, result)
    case "registerListeners": handleRegisterListenersCall(call, result)
    case "sendAttachmentText": handleSendAttachmentTextCall(call, result)
    case "login": handleLoginCall(call, result)
    case "logout": handleLogoutCall(result)
    default: result(FlutterMethodNotImplemented)
    }
  }

  init(channel: FlutterMethodChannel) {
    self.channel = channel

    super.init()
  }

  deinit {
    self.observation?.invalidate()

    NotificationCenter.default.removeObserver(self)
  }

  // MARK: - App Lifecycle Delegate Methods

  public func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any], fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) -> Bool {
    return Apptentive.shared.didReceiveRemoteNotification(userInfo, fetchCompletionHandler: completionHandler)
  }

  public func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
    Apptentive.shared.setRemoteNotificationDeviceToken(deviceToken)
  }

  public func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
    print("Failed to register for remote notifications with error: \(error.localizedDescription).")
  }

  public func applicationDidBecomeActive(_ application: UIApplication) {
    print("Application became active")
  }

  public func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
    let _ = Apptentive.shared.didReceveUserNotificationResponse(response, withCompletionHandler: completionHandler)
  }

  public func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
    let _ = Apptentive.shared.willPresent(notification, withCompletionHandler: completionHandler)
  }

  // MARK: - Apptentive Plugin Methods

  // Register the Apptentive iOS SDK
  private func handleRegisterCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let callArguments = call.arguments as? [String: Any] else {
      return result(FlutterError.init(code: Self.errorCode, message: "Expected array of strings for arguments.", details: nil))
    }

    // Get and set distribution information
    guard let distributionName = callArguments["distributionName"] as? String,
          let distributionVersion = callArguments["distributionVersion"] as? String
    else {
      return result(FlutterError.init(code: Self.errorCode, message: "Internal Apptentive Error: Missing distribution information", details: nil))
    }

    Apptentive.shared.distributionName = distributionName
    Apptentive.shared.distributionVersion = distributionVersion

    guard let (logLevel, appCredentials, apiBaseURL) = self.unpackConfiguration(callArguments["configuration"]) else {
      return result(FlutterError.init(code: Self.errorCode, message: "Missing or invalid app credentials (key/signature)", details: "Configuration is \(callArguments["configuration"] ?? "missing")"))
    }

    ApptentiveLogger.logLevel = logLevel
    let region = apiBaseURL.flatMap { ApptentiveKit.Apptentive.Region(apiBaseURL: $0) } ?? .us

    // Register Apptentive using credentials
    Apptentive.shared.register(with: appCredentials, region:region, completion: { (completionResult) -> Void in
      switch completionResult {
      case .success:
          result(true)
      case .failure(let error):
        result(FlutterError.init(code: Self.errorCode, message: "Apptentive SDK failed to register.", details: error.localizedDescription))
      }
    })
  }

  // Engage an Apptentive event with even_name, launching any valid interactions tied to the event
  private func handleEngageCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let callArguments = call.arguments as? [String: String], let eventName = callArguments["event_name"] else {
      return result(FlutterError.init(code: Self.errorCode, message: "Expected String for event name.", details: nil))
    }
    Apptentive.shared.engage(event: .init(name: eventName), from: nil, completion: { (completionResult) -> Void in
        switch completionResult {
        case .success(let success):
            result(success)
        case .failure(let error):
          result(FlutterError.init(code: Self.errorCode, message: "Apptentive event \(eventName) failed to engage.", details: error.localizedDescription))
        }
    })
  }

  // Show Message Center
  private func handleShowMessageCenter(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    Apptentive.shared.presentMessageCenter(from: nil, completion: { (completionResult) -> Void in
        switch completionResult {
        case .success(let launchedMessageCenter):
            result(launchedMessageCenter)
        case .failure(let error):
          result(FlutterError.init(code: Self.errorCode, message: "Message Center failed to present.", details: error.localizedDescription))
        }
    })
  }

  // Get the number of unread messages in Message Center
  private func handleGetUnreadMessageCount(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    result(Apptentive.shared.unreadMessageCount)
  }

  // Set person name
  private func handleSetPersonNameCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let callArguments = call.arguments as? [String: String] else {
      return result(FlutterError.init(code: Self.errorCode, message: "Expected String for person name.", details: nil))
    }

    Apptentive.shared.personName = callArguments["name"]
    result(true)
  }

  // Set person email
  private func handleSetPersonEmailCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let callArguments = call.arguments as? [String: String] else {
      return result(FlutterError.init(code: Self.errorCode, message: "Expected String for person email.", details: nil))
    }

    Apptentive.shared.personEmailAddress = callArguments["email"]
    result(true)
  }

  // Add person custom data based on key string and value of type bool, number, or string
  private func handleAddCustomPersonDataCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let (key, value) = self.convertCustomDataArguments(call.arguments) else {
      return result(FlutterError.init(code: Self.errorCode, message: "Expected String for person custom data key; Bool, Int, Double, Float or String for value.", details: nil))
    }

    Apptentive.shared.personCustomData[key] = value
    result(true)
  }

  // Remove person custom data based on key string
  private func handleRemoveCustomPersonDataCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let callArguments = call.arguments as? [String: String], let key = callArguments["key"] else {
      return result(FlutterError.init(code: Self.errorCode, message: "Expected String for person custom data key to remove.", details: nil))
    }

    Apptentive.shared.personCustomData[key] = nil
    result(true)
  }

  // Add device custom data based on key string and value of type bool, number, or string
  private func handleAddCustomDeviceDataCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let (key, value) = self.convertCustomDataArguments(call.arguments) else {
      return result(FlutterError.init(code: Self.errorCode, message: "Expected String for device custom data key; Bool, Int, Double, Float or String for value.", details: nil))
    }

    Apptentive.shared.deviceCustomData[key] = value
    result(true)
  }

  // Remove device custom data based on key string
  private func handleRemoveCustomDeviceDataCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let callArguments = call.arguments as? [String: String], let key = callArguments["key"] else {
      return result(FlutterError.init(code: Self.errorCode, message: "Expected String for custom data key to remove.", details: nil))
    }

    Apptentive.shared.deviceCustomData[key] = nil
    result(true)
  }

  // Can if an event can trigger an interaction
  private func handleCanShowInteractionCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let callArguments = call.arguments as? [String: String], let eventName = callArguments["event_name"] else {
      return result(FlutterError.init(code: Self.errorCode, message: "Expected String for event name.", details: nil))
    }

    Apptentive.shared.canShowInteraction(event: .init(name: eventName), completion: { (completionResult) -> Void in
        switch completionResult {
        case .success(let success):
            result(success)
        case .failure(let error):
          result(FlutterError.init(code: Self.errorCode, message: "Failed to check event \(eventName)", details: error.localizedDescription))
        }
    })
  }

  private func handleCanShowMessageCenterCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    print("Can Show Message Center method is not implemented.")
    result(true)
  }

  private func handleSetPushNotificationIntegrationCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    print("Push notification integration set automatically via FlutterApplicationLifeCycleDelegate.")

    UNUserNotificationCenter.current().getNotificationSettings { settings in
      switch settings.authorizationStatus {
      case .notDetermined:
        UNUserNotificationCenter.current().delegate = self
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) { granted, error in
          if granted {
            DispatchQueue.main.async {
              UIApplication.shared.registerForRemoteNotifications()
            }
          } else {
            print("Push authorization was not granted: \(error?.localizedDescription ?? "no error").")
          }
        }

      case .authorized:
        UIApplication.shared.registerForRemoteNotifications()

      default:
        print("Push authorization status is not authorized.")
        break
      }
    }
  }

  // Register Apptentive listeners
  private func handleRegisterListenersCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    self.observation = Apptentive.shared.observe(\.unreadMessageCount, options: [.new]) { _, change in
      self.channel.invokeMethod("onUnreadMessageCountChanged", arguments: ["count": change.newValue])
    }

    NotificationCenter.default.addObserver(self, selector: #selector(eventEngaged(notification:)), name: Notification.Name.apptentiveEventEngaged, object: nil)
  }

  private func handleSendAttachmentTextCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
      guard let callArguments = call.arguments as? [String: String], let message = callArguments["message"] else {
        return result(FlutterError.init(code: Self.errorCode, message: "Expected String for hidden message body.", details: nil))
      }

      Apptentive.shared.sendAttachment(message)
      result(true)
  }

  private func handleLoginCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let callArguments = call.arguments as? [String: String], let token = callArguments["token"] else {
      return result(FlutterError.init(code: Self.errorCode, message: "Expected String for token.", details: nil))
    }
    Apptentive.shared.logIn(with: token,
        completion: { (completionResult) -> Void in
              switch completionResult {
              case .success:
                  result(true)
              case .failure(let error):
                result(FlutterError.init(code: Self.errorCode, message: "Apptentive SDK failed to login.", details: error.localizedDescription))
              }
          })
    result(true)
  }

  private func handleLogoutCall(_ result: @escaping FlutterResult) {
    Apptentive.shared.logOut()
    result(true)
  }


  @objc func eventEngaged(notification: Notification) {
    guard let userInfo = notification.userInfo as? [String: String],
          let event = userInfo["eventType"]
    else {
      print("Invalid user info on event engaged notification.")
      return
    }

    switch event {
    case "com.apptentive#Survey#submit":
      self.channel.invokeMethod("onSurveyFinished", arguments: ["completed": true])

    case "com.apptentive#Survey#cancel", "com.apptentive#Survey#cancel_partial":
      self.channel.invokeMethod("onSurveyFinished", arguments: ["completed": false])

    case "com.apptentive#MessageCenter#sent":
      let sentByUser = true // TODO: get this info from event somehow.
      self.channel.invokeMethod("onMessageSent", arguments: ["sentByUser": sentByUser])

    default:
      break

    }
  }

  // MARK: - Utility Methods

  // Set ApptentiveLogger log level

  /// Sets the log level for the Apptentive SDK.
  /// - Parameter logLevel: The severity of messages that will be logged.
  private func convertLogLevel(logLevel: String?) -> LogLevel {
    switch logLevel {
    case "LogLevel.verbose": return .debug
    case "LogLevel.debug": return .info
    case "LogLevel.info": return .notice
    case "LogLevel.warn": return .warning
    case "LogLevel.error": return .critical
    case .none:
      print("")
      return .notice
    case .some(let unknownLevel):
      print("Apptentive Unknown log level: \(unknownLevel). Using .notice by default.")
      return .notice
    }
  }

  private func unpackConfiguration(_ configuration: Any?) -> (LogLevel, Apptentive.AppCredentials, URL?)? {
    guard let configuration = configuration as? [String: Any],
          let key = configuration["key"] as? String,
          let signature = configuration["signature"] as? String
    else {
      print("Missing App Credentials (key/signature) in configuration!")
      return nil
    }

    let logLevel = configuration["log_level"] as? String
    let apiBaseURL = (configuration["api_base_url"] as? String).flatMap { URL(string: $0) }
    return (self.convertLogLevel(logLevel: logLevel), .init(key: key, signature: signature), apiBaseURL)
  }

  private func convertCustomDataArguments(_ callArguments: Any?) -> (String, CustomDataCompatible)? {
    guard let callArguments = callArguments as? [String: Any],
          let value = callArguments["value"],
          let key = callArguments["key"] as? String else {
      return nil
    }

    switch value {
    case let bool as Bool:
      return (key, bool)
    case let int as Int:
      return (key, int)
    case let double as Double:
      return (key, double)
    case let float as Float:
      return (key, float)
    case let string as String:
      return (key, string)
    default:
      return nil
    }
  }
}
